dll/win32/shell32/folders/CRegFolder.cpp

```diff
@@ -861,8 +828,79 @@ HRESULT WINAPI CRegFolder::MapColumnToSCID(UINT column, SHCOLUMNID *pscid)
     return E_NOTIMPL;
 }
 
+static HRESULT CALLBACK RegFolderContextMenuCallback(IShellFolder *psf, HWND hwnd, IDataObject *pdtobj,
+                                                     UINT uMsg, WPARAM wParam, LPARAM lParam)
+{
+    if (uMsg != DFM_INVOKECOMMAND || wParam != DFM_CMD_PROPERTIES)
+        return SHELL32_DefaultContextMenuCallBack(psf, pdtobj, uMsg);
+
+    PIDLIST_ABSOLUTE pidlFolder;
+    PUITEMID_CHILD *apidl;
+    UINT cidl;
+    HRESULT hr = SH_GetApidlFromDataObject(pdtobj, &pidlFolder, &apidl, &cidl);
+    if (FAILED_UNEXPECTEDLY(hr))
+        return hr;
+
+    CRegFolder *pRegFolder = static_cast<CRegFolder*>(psf);
+    const REQUIREDREGITEM* pRequired = pRegFolder->IsRequiredItem(apidl[0]);
+    if (pRequired && pRequired->pszCpl)
+    {
+        WCHAR buf[MAX_PATH];
+        wsprintfW(buf, L"%hs", const_cast<LPCSTR>(pRequired->pszCpl));
+        hr = SHELL_ExecuteControlPanelCPL(hwnd, buf) ? S_OK : E_FAIL;
+    }
+#if 0 // Should never happen, CDesktopFolder.cpp handles this
+    else if (_ILIsDesktop(pidlFolder) && _ILIsDesktop(apidl[0]))
+    {
+        hr = SHELL_ExecuteControlPanelCPL(hwnd, L"desk.cpl") ? S_OK : E_FAIL;
+    }
+#endif
+    else if (_ILIsDesktop(pidlFolder) && _ILIsBitBucket(apidl[0]))
+    {
+        // FIXME: Use SHOpenPropSheet on its PropertySheetHandlers from the registry
+        hr = SH_ShowRecycleBinProperties(L'C') ? S_OK : E_FAIL;
+    }
+    else
+    {
+        hr = S_FALSE; // Tell the caller to run the default action
+    }
+
+    SHFree(pidlFolder);
+    _ILFreeaPidl(apidl, cidl);
+    return hr;
+}
+
+static HRESULT CRegItemContextMenu_CreateInstance(PCIDLIST_ABSOLUTE pidlFolder, HWND hwnd, UINT cidl,
+                                                  PCUITEMID_CHILD_ARRAY apidl, IShellFolder *psf, IContextMenu **ppcm)
+{
+    HKEY hKeys[3];
+    UINT cKeys = 0;
+
+    const GUID *pGuid = _ILGetGUIDPointer(apidl[0]);
+    if (pGuid)
+    {
+        WCHAR key[sizeof("CLSID\\") + 38];
+        wcscpy(key, L"CLSID\\");
+        StringFromGUID2(*pGuid, &key[6], 39);
+        AddClassKeyToArray(key, hKeys, &cKeys);
+    }
+
+    // FIXME: CRegFolder should be aggregated by its outer folder and should
+    // provide the attributes for all required non-registry folders.
+    // It currently does not so we have to ask the outer folder ourself so
+    // that we get the correct attributes for My Computer etc.
+    CComPtr<IShellFolder> pOuterSF;
+    SHBindToObject(NULL, pidlFolder, IID_PPV_ARG(IShellFolder, &pOuterSF));
+
+    SFGAOF att = (psf && cidl) ? SHGetAttributes(pOuterSF ? pOuterSF.p : psf, apidl[0], SFGAO_FOLDER) : 0;
+    if ((att & SFGAO_FOLDER) && (!pGuid || !HasCLSIDShellFolderValue(*pGuid, L"HideFolderVerbs")))
+        AddClassKeyToArray(L"Folder", hKeys, &cKeys);
+
+    return CDefFolderMenu_Create2(pidlFolder, hwnd, cidl, apidl, psf, RegFolderContextMenuCallback, cKeys, hKeys, ppcm);
+}
+
 /* In latest windows version this is exported but it takes different arguments! */
```

Comment:
This comment is interesting. Do you know what these arguments may be?

Reply:
I don't know what they are supposed to be nor does it seem to be exported anywhere?

Reply:
I think this maybe changed for Vista. If it has always been COM aggregated I suspect it takes more arguments on NT5 (but is never exported?). According to Geoff Chappell, [IRegItemFolder](https://www.geoffchappell.com/studies/windows/shell/shell32/interfaces/iregitemfolder.htm) was added to `CLSID_RegFolder` in Vista. This interface has just one method: `HRESULT Initialize (REGITEMINFO const *);`

 > ... in which REGITEMINFO is an invented typedef for an anonymous structure in the interface definition. 

That Nirsoft list also has `CRegFolder_CreateAndInit` which might combine the two again in Windows 10?